### PR TITLE
Efficient implementation of StreamEmbedTransformer

### LINKED ISSUE #2557: Blockwise unembed in StreamEmbedTransformer yield fractured kernel calls
### Describe the task. It can be a feature, documentation, etc.

In the StreamEmbedTransformer, the unembedding layers can be configured to apply Layernorm and Linear layers to the embedding output per channel by setting unembed_mode="block", see https://github.com/ecmwf/WeatherGenerator/blob/10f838d82aedb409caf96e35014aa813a73c4a02/src/weathergen/model/embeddings.py#L90-L101

This leads to many small kernels being called, see the following trace:

Image

Goals of this task:
1. Compare performance and accuracy of unembed_mode="full" vs unembed_mode="block".
2. Investigate optimizations of the block-wise unembed (e.g., vectorization).

@javak87 has started working on this.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [ ] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [x] performance

### PR DESCRIPTION
**Pull request by javak87**
## Description
Due to the many kernel calls inside the forward function of the StreamEmbedTransformer module, it needs to be optimized. The proposed implementation is slightly more optimized.

By running this:
`../WeatherGenerator-private/hpc/launch-slurm.py --time 120 --nodes=1 --base-config ./config/config_operan_georing_avhrr_forecasting_lowres.yml`

**Baseline (develop branch):**
Run_id: gtj55ov5
Number of ingested samples: 6496
Screenshot from 2026-06-30 09-42-03

**javad/dev/optimized_StreamEmbedTransformer_2557 branch**
Run_id: b04nu5cr
Number of ingested samples: 6896

Screenshot from 2026-06-30 09-41-52

The results show that performance increased by 6%.

## Issue Number
Closes #2557 

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
-   [x] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

**Comment by florianscheidl** [comment]
Hey, thanks @javak87. Have you profiled the run and seen what the kernel calls look like?

**Comment by javak87** [comment]
> Hey, thanks @javak87. Have you profiled the run and seen what the kernel calls look like?

What I did is decreasing the number of kernel calls, but I didn't look at the kernel calls.

File: src/weathergen/model/embeddings.py [hunk 1]
```diff
@@ -113,11 +113,25 @@ def forward(self, x_in):
...
-            ]
-            out = torch.stack(out, dim=1).flatten(-2, -1)
+            b, c, d = x.shape
+
+            ln_w = torch.stack([ln.weight for ln in self.ln_final])
+            ln_b = torch.stack([ln.bias for ln in self.ln_final])
+            w = torch.stack([ue.weight for ue in self.unembed])
+            ue_bias = torch.stack([ue.bias for ue in self.unembed])
+
+            x_normed = torch.nn.functional.layer_norm(
```
Comment (florianscheidl) [inline src/weathergen/model/embeddings.py:123]: Shall we use norm instead of torch.nn.functional.layer_norm here?

File: src/weathergen/model/embeddings.py [hunk 2]
```diff
@@ -113,11 +113,25 @@ def forward(self, x_in):
...
-                ue(ln(x[:, i]))
-                for i, (ue, ln) in enumerate(zip(self.unembed, self.ln_final, strict=True))
-            ]
-            out = torch.stack(out, dim=1).flatten(-2, -1)
+            b, c, d = x.shape
+
+            ln_w = torch.stack([ln.weight for ln in self.ln_final])
+            ln_b = torch.stack([ln.bias for ln in self.ln_final])
+            w = torch.stack([ue.weight for ue in self.unembed])
+            ue_bias = torch.stack([ue.bias for ue in self.unembed])
```
Comment (florianscheidl) [inline src/weathergen/model/embeddings.py:121]: Shall we move this to init?

File: src/weathergen/model/embeddings.py [hunk 3]
```diff
@@ -113,11 +113,25 @@ def forward(self, x_in):
...
+
+            x_normed = torch.nn.functional.layer_norm(
+                x.reshape(b * c, d), [d], weight=None, bias=None
+            )
+            x_normed = x_normed.reshape(b, c, d) * ln_w + ln_b
+
+            out = torch.bmm(
+                x_normed.transpose(0, 1),
+                w.transpose(1, 2),
+            ).transpose(0, 1)
```
Comment (florianscheidl) [inline src/weathergen/model/embeddings.py:131]: Can we avoid the transpose operations by indexing differently beforehand?

**Review by florianscheidl — COMMENTED** [review]
Nice! Can you check if we can optimize the indexing to avoid extra transpositions?
